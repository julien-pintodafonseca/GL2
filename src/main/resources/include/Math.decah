// Grenoble INP - Ensimag
// Projet GÃ©nie Logiciel 2020, Equipe GL2
// Library for class Math of Deca, coded in Deca

class Math {
    float MIN_VALUE = 1.4 * pow(10,-45);
    int MAX_EXPONENT = 127;
    int MIN_EXPONENT = -126;

    float ulp(float f) {
        int exp = getExponent(f);

        if (exp == MAX_EXPONENT+1) {
            return abs(f);
        } else if (exp == MIN_EXPONENT+1) {
            return MIN_VALUE;
        } else {
            println("ulp(f) not yet implemented");
        }
    }

    float sin(float f) {
        println("sin(f) not yet implemented");
    }

    float cos(float f) {
        println("cos(f) not yet implemented");
    }

    float asin(float f) {
        println("asin(f) not yet implemented");
    }

    float atan(float f) {
        println("atan(f) not yet implemented");
    }

    int getExponent(float x) {
        int e = 0;
        if (x == 0) {
            return 0;
        } else {
            while (pow(2,e) <= x) {
                e = e+1;
            }
            return e-1;
        }
    }

    float abs(float x) {
        if (x < 0) {
            return -x;
        } else {
            return x;
        }
    }

    float pow(float x, int e) {
        float p;
        if (e == 0) {
            return 1;
        } else {
            p = pow(x, e/2);

            if (e%2 == 0) {
                return p*p;
            } else if (e > 0) {
                return x*(p*p);
            } else {
                return (p*p)/x;
            }
        }
    }
}

// End of Deca Math library
