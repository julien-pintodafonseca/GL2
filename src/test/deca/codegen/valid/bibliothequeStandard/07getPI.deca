#include "Math.decah"

{
    Math m = new Math();

    println(m.getPI());     // 3.141592+00
}
