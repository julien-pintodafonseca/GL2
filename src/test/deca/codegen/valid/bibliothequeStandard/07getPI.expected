3.14159e+00
